src/openms/include/OpenMS/FILTERING/CALIBRATION/InternalCalibration.h

```diff
@@ -70,402 +70,362 @@ namespace OpenMS
-          --rt_iter;
-        }
-        // now find closest peak
-        typename MSSpectrum<InputPeakType>::ConstIterator mz_iter = rt_iter->MZBegin(ref_ids[p_id].getMZ());
-        //std::cout << mz_iter->getMZ() <<" "<<(double)ref_ids[p_id].getMZ()<<"\t";
-        double dist = ref_ids[p_id].getMZ() - mz_iter->getMZ();
-        //std::cout << dist << "\t";
-        if ((mz_iter + 1) != rt_iter->end()
-           && fabs((mz_iter + 1)->getMZ() - ref_ids[p_id].getMZ()) < fabs(dist)
-           && mz_iter != rt_iter->begin()
-           && fabs((mz_iter - 1)->getMZ() - ref_ids[p_id].getMZ()) < fabs((mz_iter + 1)->getMZ() - ref_ids[p_id].getMZ()))  // if mz_iter +1 has smaller dist than mz_iter and mz_iter-1
+        std::cout << "Input is empty." << std::endl;
+        return;
+      }
+
+      if (exp[0].getType() != SpectrumSettings::PEAKS)
+      {
+        std::cout << "Attention: this function is assuming peak data." << std::endl;
+      }
+      // check if the ids contain meta information about the peak positions
+      checkReferenceIds_(ref_ids);
+
+      std::vector<double> theoretical_masses, observed_masses;
+      for (Size p_id = 0; p_id < ref_ids.size(); ++p_id)
+      {
+        for (Size p_h = 0; p_h < ref_ids[p_id].getHits().size(); ++p_h)
         {
-          if ((use_ppm &&
-               fabs((mz_iter + 1)->getMZ() - ref_ids[p_id].getMZ()) / ref_ids[p_id].getMZ() * 1e06 < mz_tolerance) ||
-              (!use_ppm && fabs((mz_iter + 1)->getMZ() - ref_ids[p_id].getMZ()) < mz_tolerance))
+          Int charge = ref_ids[p_id].getHits()[p_h].getCharge();
+          double theo_mass = ref_ids[p_id].getHits()[p_h].getSequence().getMonoWeight(Residue::Full, charge) / (double)charge;
+          // first find corresponding ms1-spectrum
+          typename MSExperiment<InputPeakType>::ConstIterator rt_iter = exp.RTBegin(ref_ids[p_id].getRT());
+          while (rt_iter != exp.begin() && rt_iter->getMSLevel() != 1)
           {
-            //std::cout <<(mz_iter +1)->getMZ() - ref_ids[p_id].getMZ()<<"\t";
-            observed_masses.push_back((mz_iter + 1)->getMZ());
-            theoretical_masses.push_back(theo_mass);
-            //std::cout << (mz_iter +1)->getMZ() << " ~ "<<theo_mass << " charge: "<<ref_ids[p_id].getHits()[p_h].getCharge()
-            //<< "\tplus 1"<< std::endl;
+            --rt_iter;
           }
-        }
-        else if (mz_iter != rt_iter->begin()
-                && fabs((mz_iter - 1)->getMZ() - ref_ids[p_id].getMZ()) < fabs(dist))                        // if mz_iter-1 has smaller dist than mz_iter
-        {
-          if ((use_ppm &&
-               fabs((mz_iter - 1)->getMZ() - ref_ids[p_id].getMZ()) / ref_ids[p_id].getMZ() * 1e06 < mz_tolerance) ||
-              (!use_ppm && fabs((mz_iter - 1)->getMZ() - ref_ids[p_id].getMZ()) < mz_tolerance))
+          // now find closest peak
+          typename MSSpectrum<InputPeakType>::ConstIterator mz_iter = rt_iter->MZBegin(ref_ids[p_id].getMZ());
```

Comment:
I think we now have a findNearestPeak (maybe even one that takes a tolerance window)